src/Tools/dotnet-monitor/CollectionRules/Actions/CollectionRuleActionBase.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions
+{
+    internal abstract class CollectionRuleActionBase<TOptions> :
+        ICollectionRuleAction,
+        IAsyncDisposable
+    {
+        private readonly CancellationTokenSource _disposalTokenSource = new();
+
+        private Task<CollectionRuleActionResult> _completionTask;
+        private bool _isDisposed;
+
+        protected IEndpointInfo EndpointInfo { get; }
+
+        protected TOptions Options { get; }
+
+        protected CollectionRuleActionBase(IEndpointInfo endpointInfo, TOptions options)
+        {
+            // TODO: Allow null endpointInfo to allow tests to pass, but this should be provided by
+            // tests since it will be required by all aspects in the future. For example, the ActionListExecutor
+            // (which uses null in tests) will require this when needing to get process information for
+            // the actions property bag used for token replacement.
+            //EndpointInfo = endpointInfo ?? throw new ArgumentNullException(nameof(endpointInfo));
+            EndpointInfo = endpointInfo;
+            Options = options ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public async ValueTask DisposeAsync()
```

Comment:
nit: Not something we have to do now, but typically you would have another template method here to optionally override if derived classes needed to also dispose something.